// @snap/camera-kit/lib/generated-proto/pb_schema/camera_kit/v3/features/ranking_info.js
/* eslint-disable */
import Long from "long";
import _m0 from "protobufjs/minimal";
export const protobufPackage = "com.snap.camerakit.v3.features";
function createBaseRankingInfo() {
    return { rankingRequestId: "" };
}
export const RankingInfo = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.rankingRequestId !== "") {
            writer.uint32(10).string(message.rankingRequestId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRankingInfo();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.rankingRequestId = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseRankingInfo();
        message.rankingRequestId = (_a = object.rankingRequestId) !== null && _a !== void 0 ? _a : "";
        return message;
    },
};
if (_m0.util.Long !== Long) {
    _m0.util.Long = Long;
    _m0.configure();
}
