// @snap/camera-kit/lib/generated-proto/pb_schema/lenses/geocircle.js
import Long from "long";
import _m0 from "protobufjs/minimal";
import { Geopoint } from "./geopoint";
export const protobufPackage = "snapchat.lenses";
function createBaseGeocircle() {
    return { radius: 0, center: undefined };
}
export const Geocircle = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.radius !== 0) {
            writer.uint32(9).double(message.radius);
        }
        if (message.center !== undefined) {
            Geopoint.encode(message.center, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGeocircle();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.radius = reader.double();
                    break;
                case 2:
                    message.center = Geopoint.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseGeocircle();
        message.radius = (_a = object.radius) !== null && _a !== void 0 ? _a : 0;
        message.center =
            object.center !== undefined && object.center !== null ? Geopoint.fromPartial(object.center) : undefined;
        return message;
    },
};
if (_m0.util.Long !== Long) {
    _m0.util.Long = Long;
    _m0.configure();
}
