// @snap/camera-kit/lib/generated-proto/pb_schema/google/api/http.js
/* eslint-disable */
import Long from "long";
import _m0 from "protobufjs/minimal";
export const protobufPackage = "google.api";
function createBaseHttp() {
    return { rules: [], fullyDecodeReservedExpansion: false };
}
export const Http = {
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.rules) {
            HttpRule.encode(v, writer.uint32(10).fork()).ldelim();
        }
        if (message.fullyDecodeReservedExpansion === true) {
            writer.uint32(16).bool(message.fullyDecodeReservedExpansion);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseHttp();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.rules.push(HttpRule.decode(reader, reader.uint32()));
                    break;
                case 2:
                    message.fullyDecodeReservedExpansion = reader.bool();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            rules: Array.isArray(object === null || object === void 0 ? void 0 : object.rules) ? object.rules.map((e) => HttpRule.fromJSON(e)) : [],
            fullyDecodeReservedExpansion: isSet(object.fullyDecodeReservedExpansion)
                ? Boolean(object.fullyDecodeReservedExpansion)
                : false,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.rules) {
            obj.rules = message.rules.map((e) => (e ? HttpRule.toJSON(e) : undefined));
        }
        else {
            obj.rules = [];
        }
        message.fullyDecodeReservedExpansion !== undefined &&
            (obj.fullyDecodeReservedExpansion = message.fullyDecodeReservedExpansion);
        return obj;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseHttp();
        message.rules = ((_a = object.rules) === null || _a === void 0 ? void 0 : _a.map((e) => HttpRule.fromPartial(e))) || [];
        message.fullyDecodeReservedExpansion = (_b = object.fullyDecodeReservedExpansion) !== null && _b !== void 0 ? _b : false;
        return message;
    },
};
function createBaseHttpRule() {
    return { selector: "", pattern: undefined, body: "", additionalBindings: [] };
}
export const HttpRule = {
    encode(message, writer = _m0.Writer.create()) {
        var _a, _b, _c, _d, _e, _f;
        if (message.selector !== "") {
            writer.uint32(10).string(message.selector);
        }
        if (((_a = message.pattern) === null || _a === void 0 ? void 0 : _a.$case) === "get") {
            writer.uint32(18).string(message.pattern.get);
        }
        if (((_b = message.pattern) === null || _b === void 0 ? void 0 : _b.$case) === "put") {
            writer.uint32(26).string(message.pattern.put);
        }
        if (((_c = message.pattern) === null || _c === void 0 ? void 0 : _c.$case) === "post") {
            writer.uint32(34).string(message.pattern.post);
        }
        if (((_d = message.pattern) === null || _d === void 0 ? void 0 : _d.$case) === "delete") {
            writer.uint32(42).string(message.pattern.delete);
        }
        if (((_e = message.pattern) === null || _e === void 0 ? void 0 : _e.$case) === "patch") {
            writer.uint32(50).string(message.pattern.patch);
        }
        if (((_f = message.pattern) === null || _f === void 0 ? void 0 : _f.$case) === "custom") {
            CustomHttpPattern.encode(message.pattern.custom, writer.uint32(66).fork()).ldelim();
        }
        if (message.body !== "") {
            writer.uint32(58).string(message.body);
        }
        for (const v of message.additionalBindings) {
            HttpRule.encode(v, writer.uint32(90).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseHttpRule();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.selector = reader.string();
                    break;
                case 2:
                    message.pattern = { $case: "get", get: reader.string() };
                    break;
                case 3:
                    message.pattern = { $case: "put", put: reader.string() };
                    break;
                case 4:
                    message.pattern = { $case: "post", post: reader.string() };
                    break;
                case 5:
                    message.pattern = { $case: "delete", delete: reader.string() };
                    break;
                case 6:
                    message.pattern = { $case: "patch", patch: reader.string() };
                    break;
                case 8:
                    message.pattern = { $case: "custom", custom: CustomHttpPattern.decode(reader, reader.uint32()) };
                    break;
                case 7:
                    message.body = reader.string();
                    break;
                case 11:
                    message.additionalBindings.push(HttpRule.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            selector: isSet(object.selector) ? String(object.selector) : "",
            pattern: isSet(object.get)
                ? { $case: "get", get: String(object.get) }
                : isSet(object.put)
                    ? { $case: "put", put: String(object.put) }
                    : isSet(object.post)
                        ? { $case: "post", post: String(object.post) }
                        : isSet(object.delete)
                            ? { $case: "delete", delete: String(object.delete) }
                            : isSet(object.patch)
                                ? { $case: "patch", patch: String(object.patch) }
                                : isSet(object.custom)
                                    ? { $case: "custom", custom: CustomHttpPattern.fromJSON(object.custom) }
                                    : undefined,
            body: isSet(object.body) ? String(object.body) : "",
            additionalBindings: Array.isArray(object === null || object === void 0 ? void 0 : object.additionalBindings)
                ? object.additionalBindings.map((e) => HttpRule.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o;
        const obj = {};
        message.selector !== undefined && (obj.selector = message.selector);
        ((_a = message.pattern) === null || _a === void 0 ? void 0 : _a.$case) === "get" && (obj.get = (_b = message.pattern) === null || _b === void 0 ? void 0 : _b.get);
        ((_c = message.pattern) === null || _c === void 0 ? void 0 : _c.$case) === "put" && (obj.put = (_d = message.pattern) === null || _d === void 0 ? void 0 : _d.put);
        ((_e = message.pattern) === null || _e === void 0 ? void 0 : _e.$case) === "post" && (obj.post = (_f = message.pattern) === null || _f === void 0 ? void 0 : _f.post);
        ((_g = message.pattern) === null || _g === void 0 ? void 0 : _g.$case) === "delete" && (obj.delete = (_h = message.pattern) === null || _h === void 0 ? void 0 : _h.delete);
        ((_j = message.pattern) === null || _j === void 0 ? void 0 : _j.$case) === "patch" && (obj.patch = (_k = message.pattern) === null || _k === void 0 ? void 0 : _k.patch);
        ((_l = message.pattern) === null || _l === void 0 ? void 0 : _l.$case) === "custom" &&
            (obj.custom = ((_m = message.pattern) === null || _m === void 0 ? void 0 : _m.custom) ? CustomHttpPattern.toJSON((_o = message.pattern) === null || _o === void 0 ? void 0 : _o.custom) : undefined);
        message.body !== undefined && (obj.body = message.body);
        if (message.additionalBindings) {
            obj.additionalBindings = message.additionalBindings.map((e) => (e ? HttpRule.toJSON(e) : undefined));
        }
        else {
            obj.additionalBindings = [];
        }
        return obj;
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w;
        const message = createBaseHttpRule();
        message.selector = (_a = object.selector) !== null && _a !== void 0 ? _a : "";
        if (((_b = object.pattern) === null || _b === void 0 ? void 0 : _b.$case) === "get" && ((_c = object.pattern) === null || _c === void 0 ? void 0 : _c.get) !== undefined && ((_d = object.pattern) === null || _d === void 0 ? void 0 : _d.get) !== null) {
            message.pattern = { $case: "get", get: object.pattern.get };
        }
        if (((_e = object.pattern) === null || _e === void 0 ? void 0 : _e.$case) === "put" && ((_f = object.pattern) === null || _f === void 0 ? void 0 : _f.put) !== undefined && ((_g = object.pattern) === null || _g === void 0 ? void 0 : _g.put) !== null) {
            message.pattern = { $case: "put", put: object.pattern.put };
        }
        if (((_h = object.pattern) === null || _h === void 0 ? void 0 : _h.$case) === "post" && ((_j = object.pattern) === null || _j === void 0 ? void 0 : _j.post) !== undefined && ((_k = object.pattern) === null || _k === void 0 ? void 0 : _k.post) !== null) {
            message.pattern = { $case: "post", post: object.pattern.post };
        }
        if (((_l = object.pattern) === null || _l === void 0 ? void 0 : _l.$case) === "delete" && ((_m = object.pattern) === null || _m === void 0 ? void 0 : _m.delete) !== undefined && ((_o = object.pattern) === null || _o === void 0 ? void 0 : _o.delete) !== null) {
            message.pattern = { $case: "delete", delete: object.pattern.delete };
        }
        if (((_p = object.pattern) === null || _p === void 0 ? void 0 : _p.$case) === "patch" && ((_q = object.pattern) === null || _q === void 0 ? void 0 : _q.patch) !== undefined && ((_r = object.pattern) === null || _r === void 0 ? void 0 : _r.patch) !== null) {
            message.pattern = { $case: "patch", patch: object.pattern.patch };
        }
        if (((_s = object.pattern) === null || _s === void 0 ? void 0 : _s.$case) === "custom" && ((_t = object.pattern) === null || _t === void 0 ? void 0 : _t.custom) !== undefined && ((_u = object.pattern) === null || _u === void 0 ? void 0 : _u.custom) !== null) {
            message.pattern = { $case: "custom", custom: CustomHttpPattern.fromPartial(object.pattern.custom) };
        }
        message.body = (_v = object.body) !== null && _v !== void 0 ? _v : "";
        message.additionalBindings = ((_w = object.additionalBindings) === null || _w === void 0 ? void 0 : _w.map((e) => HttpRule.fromPartial(e))) || [];
        return message;
    },
};
function createBaseCustomHttpPattern() {
    return { kind: "", path: "" };
}
export const CustomHttpPattern = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.kind !== "") {
            writer.uint32(10).string(message.kind);
        }
        if (message.path !== "") {
            writer.uint32(18).string(message.path);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCustomHttpPattern();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.kind = reader.string();
                    break;
                case 2:
                    message.path = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            kind: isSet(object.kind) ? String(object.kind) : "",
            path: isSet(object.path) ? String(object.path) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        message.kind !== undefined && (obj.kind = message.kind);
        message.path !== undefined && (obj.path = message.path);
        return obj;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseCustomHttpPattern();
        message.kind = (_a = object.kind) !== null && _a !== void 0 ? _a : "";
        message.path = (_b = object.path) !== null && _b !== void 0 ? _b : "";
        return message;
    },
};
if (_m0.util.Long !== Long) {
    _m0.util.Long = Long;
    _m0.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
